[RFC] Travis: Refactor CI files, use container infrastructure.

fwalch opened the pull request:
- Split build steps to utilize the Travis build lifecycle.
- Move shell code from `.travis.yml` into Bash files in `.ci/`,
  one file for each step of the Travis build lifecycle.
- Use configuration variables in `.travis.yml` to change
  build behavior (e.g. build 32-bit with `BUILD_32BIT=ON`).
- Keep all configuration in environment variables in
  `.travis.yml`. In scripts, concatenate environment variables
  according to configuration to change to different behavior.
- Add GCC 5 Linux builds.
- Use Travis's caching feature [1] for third-party dependencies
  and pip packages.
- Allow failures MSan, as the errors it reports have to be
  fixed first.

Valgrind is still disabled, but can be enabled by setting
`env: VALGRIND=ON` for a job in `.travis.yml`.

[1] http://docs.travis-ci.com/user/caching

~~`WIP` since I couldn't test the OS X builds, so they might not work currently.~~ _Edit_: Updated PR description with new commit message.
Also, since we don't have `sudo`, we cannot create new groups for the tests:

```
# Adds user to a dummy group.
# That allows to test changing the group of the file by `os_fchown`.
```

ZyX-I commented:
I am wondering whether this will allow re-enabling valgrind in tests. Various caches should cut off the installation time and containers also waste less resources on themselves.

fwalch commented:
@ZyX-I Good question, let's re-enable it and see what happens.

fwalch commented:
One note about the caches: The Travis docu says that PRs currently use caches of the branch they're supposed to be merged into. This means that a PR that changes the dependencies could break other PR's builds. To try to prevent this, I added a commit to only update the third-party dependencies cache if the build was successful (i.e. all tests passed).

fwalch commented:
The Valgrind builds still take 40min.. I think that's too long to run on every PR. But we could set up "nightly builds" with `bot-ci` or the QuickBuild server.

fwalch commented:
> The Travis docu says that PRs currently use caches of the branch they're supposed to be merged into.

Looking at the build output, this doesn't seem to hold anymore: `fetching PR.2938/cache--compiler-clang-3.6.tgz`. So I guess we don't have to worry about this :-)

fwalch committed:
CMake: Allow overriding third-party install prefix.

fwalch commented:
Okay, this seems to work now, with and without cached dependencies. Marking `RFC` to get comments.

One strange thing I noticed: `pip` is upgraded in `before_install`.. or is it?

```
# When upgrading pip:
You are using pip version 6.0.8, however version 7.1.0 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
Collecting pip from https://pypi.python.org/packages/py2.py3/p/pip/pip-7.1.0-py2.py3-none-any.whl#md5=b108384a762825ec20345bb9b5b7209f
  Using cached pip-7.1.0-py2.py3-none-any.whl
Installing collected packages: pip
Successfully installed pip-6.0.8
# Next invocation of pip:
You are using pip version 6.0.8, however version 7.1.0 is available.
You should consider upgrading via the 'pip install --upgrade pip' command.
```

Note how it says it will install 7.1.0, but then reports that it installed 6.0.8. It should be installed to `~/.local/bin`, which is in the `PATH`.

tarruda commented:
@fwalch great improvement, not only to build times but also to the scripts. LGTM

fwalch committed:
Travis: Refactor CI files, use container infrastructure.

 * Split build steps to utilize the Travis build lifecycle.
 * Move shell code from `.travis.yml` into Bash files in `.ci/`,
   one file for each step of the Travis build lifecycle.
 * Use configuration variables in `.travis.yml` to change
   build behavior (e.g. build 32-bit with `BUILD_32BIT=ON`).
 * Keep all configuration in environment variables in
   `.travis.yml`. In scripts, concatenate environment variables
   according to configuration to change to different behavior.
 * Add GCC 5 builds for Linux.
 * Use Travis's caching feature [1] for third-party dependencies
   and pip packages.
 * Allow failures MSan, as the errors it reports have to be
   fixed first.

Valgrind is still disabled, but can be enabled by setting
`env: VALGRIND=ON` for a job in `.travis.yml`.

[1] http://docs.travis-ci.com/user/caching

fwalch commented:
Rebased after #2807.

fwalch committed:
Travis: Use gcov that matches the used compiler.

This fixes gcov/coveralls warnings like the following:

    Segmentation fault (core dumped)
    charset.c.gcno:version '501*', prefer '406*'
    Out of memory allocating 33061786568 bytes after a total of 2522648 bytes

http://stackoverflow.com/a/14676272/249642

fwalch commented:
Re-introduced a job with GCC 4.6 to make sure we don't break something for older compilers. Fixed gcov/coveralls warnings (`charset.c.gcno:version '501*', prefer '406*'`) by using a `gcov` (or `llvm-cov`) matching the used compiler. Upload still fails:

```
{u'url': u'', u'message': u'Build processing error.', u'error': True}
coveralls upload failed.
```

This happens for all jobs. It also occurs with other PRs/`master` (according to the coveralls website, the last build was 17h ago, but I merged #2807 in the meantime), so maybe coveralls has some problems atm.

Still don't know why `pip` isn't upgraded (https://github.com/neovim/neovim/pull/2938#issuecomment-118535109).

fwalch committed:
Travis: Check some configuration variables in before_script.

fwalch commented on cmake/i386-linux-gnu.toolchain.cmake:
```diff
@@ -1,5 +1,7 @@
 set(CMAKE_SYSTEM_VERSION gnu)
-set(CMAKE_C_COMPILER gcc)
+if(NOT ${CMAKE_C_COMPILER})
+  set(CMAKE_C_COMPILER gcc)
```
@jszakmeister I changed this to be able to use a different compiler (in this case, `gcc-5`) to build for 32 bit. `NOT CMAKE_C_COMPILER` and `NOT DEFINED CMAKE_C_COMPILER` both didn't work. Do you think it's okay like this?

Pyrohh commented:
> Re-introduced a job with GCC 4.6 to make sure we don't break something for older compilers

Has this ever been an issue?

fwalch commented:
Used the wrong `gcov` for OS X clang, but should be fixed now. Added an additional commit to verify that `$LLVM_SYMBOLIZER` and `$GCOV` can be executed.

> Has this ever been an issue?

@Pyrohh So far we always used GCC 4.6 (for Linux, OSX used 4.9) on Travis, so it's hard to say. Probably not...

Pyrohh commented:
> @Pyrohh So far we always used GCC 4.6 (for Linux, OSX used 4.9) on Travis, so it's hard to say. Probably not...

I'd say we should drop it for now; if anything comes up it will come up quickly. Besides, GCC has a lot of backwards compatibility, so I'd be surprised if anything does break (besides some flags which might not work).

Pyrohh commented:
@fwalch Also, regarding `pip`: I'd like to see it updated too, but is there much benefit in doing it ourselves? If `6.x.x` works fine then lets just stick with that (unless there's some big speed benefit I'm unaware of).

As a side note, I saw https://github.com/travis-ci/travis-ci/issues/3968 and wonder when the "next build environment image is provisioned".

fwalch commented:
Okay, dropped the GCC 4.6 build.

Regarding pip, you're right, it's not really necessary to update it, I'm just wondering why it doesn't work.

Pyrohh commented on .ci/before_install.sh:
```diff
@@ -0,0 +1,13 @@
+if [[ "${TRAVIS_OS_NAME}" == osx ]]; then
+  brew update
+fi
+pip install --user --upgrade pip
```
Since you said it doesn't really matter if this is updated, could you remove this line? or does it work now?

fwalch replied:
It doesn't work (see also https://github.com/travis-ci/travis-ci/issues/41949, but I'd just keep it in there, maybe it will be fixed at some point.

fwalch commented:
Okay, so the `pip` upgrade problem is due to the way how Travis installs pip on their workers. As I said in https://github.com/neovim/neovim/pull/2938#discussion_r33911583, I'd still keep the upgrade in the scripts, it will probably be fixed by Travis at some point.

Noticed some small things and amended the commits earlier. I think this PR is finished now.

tarruda commented:
@justinmk can we merge this? I'd love some faster travis builds in PRs

tarruda commented:
merged

justinmk commented:
This was a great contribution, but it seems like gcov errors are now more frequent.

fwalch commented:
> This was a great contribution, but it seems like gcov errors are now more frequent.

The `gcov` errors should have been completely fixed, because now always the appropriate `gcov` executable is used. Do you mean coveralls upload errors (I have been away for a few days)? If so, then I don't think this is related to this PR, I noticed that *no* coveralls upload succeeded for a while when working on this (i.e. I think coveralls had problems with their site).

justinmk commented:
> The gcov errors should have been completely fixed, because now always the appropriate gcov executable is used. 

@fwalch  The gcov errors I am referring to are "gcda merge mismatch" messages, see this recent build for example: https://travis-ci.org/neovim/neovim/jobs/71351785

    log.c.gcda:Merge mismatch for function 0
